src/mlpack/methods/ann/layer/sigmoid_cross_entropy_error_impl.hpp

```diff
@@ -31,39 +31,43 @@ template<typename eT>
 inline double SigmoidCrossEntropyError<InputDataType, OutputDataType>::Forward(
     const arma::Mat<eT>&& input, const arma::Mat<eT>&& target)
 {
-  arma::uvec positive = arma::find(input > 0);
-  arma::mat output;
-  SoftplusFunction::Fn(static_cast<arma::mat>(-arma::abs(input)), output);
-  double loss = arma::accu(input(positive)) - arma::accu(input % target) +
-        arma::accu(output);
+  double maximum = 0;
+  for (size_t i = 0; i < input.n_elem; ++i)
+  {
+    maximum += std::max(input[i], 0.0) +
+        SoftplusFunction::Fn(-std::abs(input[i]));
+  }
+  double loss = maximum - arma::accu(input % target);
 
   return loss;
 }
 
 template<typename InputDataType, typename OutputDataType>
 template<typename eT>
-inline  void SigmoidCrossEntropyError<InputDataType, OutputDataType>::Backward(
+inline void SigmoidCrossEntropyError<InputDataType, OutputDataType>::Backward(
     const arma::Mat<eT>&& input,
     const arma::Mat<eT>&& target,
     arma::Mat<eT>&& output)
 {
-  arma::mat temp;
   output.set_size(input.n_rows, input.n_cols);
 
-  arma::uvec positive = arma::find(input > 0);
-  arma::uvec negative = arma::find(input < 0 || input == 0);
-
-  SoftplusFunction::Deriv(static_cast<arma::mat>(-arma::abs(input(positive))),
-      temp);
-  output(positive) = 1 - target(positive) - temp;
-  SoftplusFunction::Deriv(static_cast<arma::mat>(-arma::abs(input(negative))),
-      temp);
-  output(negative) = -(target(negative)) - negative;
+  for (size_t i = 0; i < output.n_elem; ++i)
+  {
+    const double val = target[i];
+    if (val > 0)
+    {
+      output[i] = 1 - val - SoftplusFunction::Deriv(-std::abs(input[i]));
```

Comment:
Not sure if this is faster, but what about using:

```c++
output = 1 - target - SoftplusFunction::Deriv(-std::abs(input));
output = (-1 * output > 0) + 1;
```

do you mind to give this is try? A simple test would use both snippets and time everything e.g. with the C++11 time class.

Reply:
That will lead us back to Kris' original implementation which was avoided to prevent the usage of a temporary matrix in the first place.

Reply:
Doing this:
```
  auto start = std::chrono::system_clock::now();
  double error1 = module.Forward(std::move(input1), std::move(target1));
  auto end = std::chrono::system_clock::now();
  std::chrono::duration<double> elapsed_seconds = end - start;
```
and adding up the `elapsed_seconds` over all the calls, I'm getting an average of `3.5e-06` seconds.

Reply:
Thanks for looking into the issue, just wanted to make sure we don't miss anything here.

Is this for the current implementation? What was the input size?

Reply:
Yes, this is for the current implementation. I timed it for only the test cases we have written, what input size would you like to know the benchmarks for?

Reply:
Thanks for the fast response, what do you think about 256, 512, 1024 and 2048? If you have time can you time the other approach as well, don't feel obligated I can also run the tests.

Reply:
@ShikharJ @kris-singh Did you compare the results of the ``Backward`` function against Tensorflow, since I don't think the derivative is correct for the case ``x <= 0`` http://www.wolframalpha.com/input/?i=d%2Fdx++max(x,+0)+-+x+*+z+%2B+log(1+%2B+exp(-abs(x)))

If you checked the backward pass, we should take a second look, if not we can simplify the overall backward function.

Reply:
I guess the derivative is correct:
Let x > 0. Then
f  ( - abs ( x ) ) = f ( - x ) and
d/dx [ f ( - abs ( x ) ) ] = d/dx [ f ( - x ) ] = - d f ( y ) / d y, where y = - x.

Let x < 0. Then
f  ( - abs ( x ) ) = f ( x ) and
d/dx [ f ( - abs ( x ) ) ] = d/dx [ f ( x ) ] =  d f ( x ) / d x

Reply:
Right, but did you check the sign of the second term, shouldn't this be:

```
z - 1/(e^x+1)
```
also the current method uses target inside the loop, where Kris original method used input.

Reply:
In this discussion `x` is `input` and `z` is `target` (unless I've misunderstood), so it seems like at the very least the conditional should be `if (input[i] > 0)` and not `if (val > 0)`.  Let me know if I am wrong about that.

But to the best of my (quick) analysis, I think both the derivative that @lozhnikov and the Wolfram Alpha results are correct; but it seems to me that the Wolfram Alpha results (specifically the first "alternate form") offers us a really nice simplification:

```
// I'm avoiding SoftplusFunction::Deriv() here so the entire expression is evaluated at once by Armadillo.
output = 1.0 / (1.0 + arma::exp(-input)) - target;
```

Like I said, let me know if I am wrong.  I am not fully confident but this is what it seems like to me.

Reply:
> the current method uses target inside the loop, where Kris original method used input.

This is my error from https://github.com/mlpack/mlpack/pull/1104/files#r137060612, sorry about that. :(

Reply:
@zoq The derivative is correct to the best of my knowledge as well. Also, we can just ignore, `SoftplusFunction::Deriv()` and go by the implementation given by @rcurtin (and yes, my bad that I didn't notice that input should be inside the loop). What made you arrive at this though?
```
z - 1/(e^x+1)
```

Reply:
I ran this snippet:
```
input1 = arma::ones(1, 256);
  target1 = arma::ones(1, 256);
  auto start = std::chrono::system_clock::now();
  error1 = module.Forward(std::move(input1), std::move(target1));
  module.Backward(std::move(input1), std::move(target1), std::move(output));
  auto end = std::chrono::system_clock::now();
  std::chrono::duration<double> elapsed_seconds = end - start;
  std::cout << elapsed_seconds.count() << std::endl;

  input1 = arma::ones(1, 512);
  target1 = arma::ones(1, 512);
  start = std::chrono::system_clock::now();
  error1 = module.Forward(std::move(input1), std::move(target1));
  module.Backward(std::move(input1), std::move(target1), std::move(output));
  end = std::chrono::system_clock::now();
  elapsed_seconds = end - start;
  std::cout << elapsed_seconds.count() << std::endl;

  input1 = arma::ones(1, 1024);
  target1 = arma::ones(1, 1024);
  start = std::chrono::system_clock::now();
  error1 = module.Forward(std::move(input1), std::move(target1));
  module.Backward(std::move(input1), std::move(target1), std::move(output));
  end = std::chrono::system_clock::now();
  elapsed_seconds = end - start;
  std::cout << elapsed_seconds.count() << std::endl;

  input1 = arma::ones(1, 2048);
  target1 = arma::ones(1, 2048);
  start = std::chrono::system_clock::now();
  error1 = module.Forward(std::move(input1), std::move(target1));
  module.Backward(std::move(input1), std::move(target1), std::move(output));
  end = std::chrono::system_clock::now();
  elapsed_seconds = end - start;
  std::cout << elapsed_seconds.count() << std::endl;
```
And got the following output for the current implementation (averaged over 5 instances):
```
2.0128e-05 (256)
5.8301e-05 (512)
9.0369e-05 (1024)
0.000166852 (2048)
```
and this on @rcurtin's approach (again, averaged over 5 instances):
```
2.4565e-05 (256)
0.0132882 (512) (weird?)
8.9329e-05 (1024)
0.000164627 (2048)
```

Reply:
@zoq It seems that for smaller instances, the single loop works better, but for larger sets, the `arma::exp()` approach might pan out just slightly better. How do you wish to proceed here?

Reply:
If there are no objections let's go with:

```c++
// I'm avoiding SoftplusFunction::Deriv() here so the entire
// expression is evaluated at once by Armadillo.
output = 1.0 / (1.0 + arma::exp(-input)) - target;
```

I think there is no need for an additional check. Also thanks for all the timings.

Reply:
Would you have any idea why I got that wild timing for the case of `512`? Was that probably because of a problem with my machine?

Reply:
hm, not sure, if you rerun the test do you get similar results?

Reply:
I'll have to build and check again. But that test was averaged over 5 instances. And the output for the case was pretty much the same each time.

Reply:
Will also run the timings and post the results here.

Reply:
Results:

```
256: 0.281 ms
512: 0.563 ms
1014: 0.979 ms
2048: 1.989 ms
```

Reply:
Ah, thanks, I rechecked again. No problems this time. Thanks, I updated the implementation here.